gcloud-java-core/src/main/java/com/google/gcloud/BaseServiceException.java

```diff
@@ -16,26 +16,80 @@
 
 package com.google.gcloud;
 
+import com.google.api.client.googleapis.json.GoogleJsonError;
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+
+import java.io.IOException;
+import java.net.SocketTimeoutException;
+import java.util.Collections;
+import java.util.Set;
+
 /**
  * Base class for all service exceptions.
  */
 public class BaseServiceException extends RuntimeException {
 
   private static final long serialVersionUID = 5028833760039966178L;
+  public static final int UNKNOWN_CODE = 0;
 
   private final int code;
   private final boolean retryable;
+  private final boolean idempotent;
+
+  public BaseServiceException(IOException exception, boolean idempotent) {
+    super(exception.getMessage(), exception);
+    this.code = statusCode(exception);
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(exception);
+  }
+
+  public BaseServiceException(GoogleJsonError error, boolean idempotent) {
+    super(error.getMessage());
+    this.code = error.getCode();
+    this.idempotent = idempotent;
+    this.retryable = idempotent && isRetryable(error);
+  }
 
-  public BaseServiceException(int code, String message, boolean retryable) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent) {
     super(message);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
   }
 
-  public BaseServiceException(int code, String message, boolean retryable, Exception cause) {
+  public BaseServiceException(int code, String message, boolean retryable, boolean idempotent,
+      Exception cause) {
     super(message, cause);
     this.code = code;
-    this.retryable = retryable;
+    this.idempotent = idempotent;
+    this.retryable = idempotent && retryable;
+  }
+
+  protected Set<Integer> retryableCodes() {
```

Comment:
Nit (and maybe not relevant if this doesn't particular PR isn't meant to be merged in), but would it be viable to make this method (and the class) abstract?